### ISSUE TITLE
UDF returning RECORD hits InvalidAssignmentCastError in SELECT ... FROM table

### ISSUE THREAD

**Issue by msirek** [cockroachdb/cockroach#96047]
The following UDF returning a RECORD works on Postgres but fails on CRDB:
```sql
CREATE TABLE t1 (a INT);
CREATE FUNCTION comp2() RETURNS RECORD IMMUTABLE LEAKPROOF LANGUAGE SQL AS 'SELECT ROW(1.1)';
EXPLAIN SELECT t1.*, comp2(), comp2() FROM t1;
(42804) value type tuple{decimal} doesn't match type tuple of column "?column?"
errors.go:61: in NewInvalidAssignmentCastError()
```

Jira issue: CRDB-23904

**Comment by mgartner** [comment]
Related to #95171.

**Comment by rharding6373** [comment]
Fixed by [#97199](https://github.com/cockroachdb/cockroach/pull/97199)
